persistence/docs/PROTOCOL_STATE_HASH.md

```diff
@@ -0,0 +1,131 @@
+| --------------------- | ------------------------------------- | ------------------------------------------------------ | --------------------------------- | ------------------- | -------------------------------------------------------------------------------- |
+| Data Tables           | SQL Database / Engine                 | MySQL, SQLite, PostgreSQL                              | PostgresSQL                       | Validator SQL Table | Validating & updating information when applying a transaction                    |
+| Merkle Trees          | Merkle Trie backed by Key-Value Store | Celestia's SMT, Libra's JMT, Cosmos' IAVL, Verkle Tree | Celestia's SMT                    | Fisherman Trie      | Maintains the state of all account based trees                                   |
+| Blocks                | Serialization Codec                   | Amino, Protobuf, Thrift, Avro                          | Protobuf                          | Block protobuf      | Serialized and inserted into the Block Store                                     |
+| Objects (e.g. Actors) | Serialization Codec                   | Amino, Protobuf, Thrift, Avro                          | Protobuf                          | Servicer protobuf   | Serialized and inserted into the corresponding Tree                              |
+| Block Store           | Key Value Store                       | LevelDB, BadgerDB, RocksDB, BoltDB                     | BadgerDb                          | Block Store         | Maintains a key-value store of the blockchain blocks                             |
+| Transaction Indexer   | Key Value Store                       | LevelDB, BadgerDB, RocksDB, BoltDB                     | BadgerDB                          | Tx Indexer          | Indexes transactions in different ways for fast queries, presence checks, etc... |
+
+### Block Proto
+
+The block protobuf that is serialized and store in the block store can be found in `persistence/proto/block_persistence.proto`. This proto contains the `stateHash` along with the corresponding height.
+
+### Trees
+
+An individual Merkle Tree is created for each type of actor, record or data type. Each of these is backed by its own key-value store.
+
+**Actor Merkle Trees**:
+
+- Applications
+- Validators
+- Fisherman
+- ServiceNodes
+
+**Account Merkle Trees**:
+
+- Accounts
+- Pools
+
+**Data Merkle Trees**
+
+- Transactions
+- Parameters
+- Flags
+
+## Compute State Hash
+
+_Note: `GetRecordsUpdatedAtHeight` is an abstraction for retrieving all the records from the corresponding SQL tables depending on the type of record (Actors, Transactions, Params, etc...)_
+
+This flow shows the interaction between the `PostgresDB` and `MerkleTrees` listed above to compute the state hash. Assuming the process of applying a proposal block to the current context (i.e. the uncommited SQL state) is done, the following steps compute the hash of the new world state.
+
+1. Loop over all of the merkle tree types
```

Comment:
What happens if a new tree type gets introduced at some point? Do we enforce ordering deterministically?

Reply:
Nevermind, answered myself below

Reply:
I'm glad you brought this up though. I added this in #361 and will add a clarifying comment here too.

```
18. Consider specifying the order of the merkle trees in a .proto enum rather than in .go iota to separate state commitment from implementation.
```